File: geekbang/algorithm-models/3.2-组合递归模板/combine.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        self.ans = []
        self.s = []
        self.n = n
        self.k = k
        self.findSubsets(1)
        return self.ans

    def findSubsets(self, index):
        # 已经选了超过k个，
        # 或者把剩下的全选上也不够k个，退出
        if len(self.s) > self.k or len(self.s) + self.n - index + 1 < self.k:
            return
        if index == self.n + 1:
            self.ans.append(self.s[:]) # make a copy
            return
        self.findSubsets(index + 1)
        self.s.append(index)
        self.findSubsets(index + 1)
        self.s.pop()

File: geekbang/algorithm-models/3.2-组合递归模板/combine.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        this->n = n;
        this->k = k;
        findSubsets(1);
        return ans;
    }


    void findSubsets(int index) {
        // 已经选了超过k个，
        // 或者把剩下的全选上也不够k个，退出
        if (s.size() > k || s.size() + n - index + 1 < k) return;
        if (index == n + 1) {
            ans.push_back(s);
            return;
        }
        findSubsets(index + 1);
        s.push_back(index);
        findSubsets(index + 1);
        s.pop_back();
    }


private:
    vector<vector<int>> ans;
    vector<int> s;
    int n;
    int k;
};
